src/main/java/org/ow2/proactive/connector/iaas/cloud/provider/vmware/VMWareProviderVirtualMachineUtil.java

```diff
@@ -64,91 +64,130 @@
-                    .max(Comparator.comparing(datastore -> (int) datastore.getSummary().getFreeSpace()))
-                    .orElseGet(null);
+    public Optional<Datastore> getDatastoreWithMostSpaceFromHost(HostSystem host) {
+        try {
+            return Lists.newArrayList(host.getDatastores())
+                        .stream()
+                        .filter(datastore -> datastore.getSummary().isAccessible())
+                        .max(Comparator.comparing(datastore -> (int) datastore.getSummary().getFreeSpace()));
+        } catch (RemoteException e) {
+            throw new RuntimeException("ERROR when retrieving VMWare datastores from host: " + host.getName(), e);
+        }
     }
 
-    public Datastore getDatastoreWithMostSpaceFromPool(ResourcePool resourcePool) throws RemoteException {
-        return Lists.newArrayList(resourcePool.getOwner().getDatastores())
-                    .stream()
-                    .filter(datastore -> datastore.getSummary().isAccessible())
-                    .max(Comparator.comparing(datastore -> (int) datastore.getSummary().getFreeSpace()))
-                    .orElseGet(null);
+    public Optional<Datastore> getDatastoreWithMostSpaceFromPool(ResourcePool resourcePool) {
+        try {
+            return Lists.newArrayList(resourcePool.getOwner().getDatastores())
+                        .stream()
+                        .filter(datastore -> datastore.getSummary().isAccessible())
+                        .max(Comparator.comparing(datastore -> (int) datastore.getSummary().getFreeSpace()));
+        } catch (RemoteException e) {
+            throw new RuntimeException("ERROR when retrieving VMWare resource pool's owner from: " +
+                                       resourcePool.getName(), e);
+        }
     }
 
-    public ResourcePool getRandomResourcePool(Folder rootFolder) throws RemoteException {
-        List<ResourcePool> resourcePools = Lists.newArrayList(new InventoryNavigator(rootFolder).searchManagedEntities(EntityType.POOL.getValue()))
-                                                .stream()
-                                                .map(resourcePool -> (ResourcePool) resourcePool)
-                                                .collect(Collectors.toCollection(ArrayList::new));
-        return resourcePools.isEmpty() ? null : resourcePools.get(new Random().nextInt(resourcePools.size()));
+    public Optional<ResourcePool> getRandomResourcePool(Folder rootFolder) {
+        List<ResourcePool> resourcePools = null;
+        try {
+            resourcePools = Lists.newArrayList(new InventoryNavigator(rootFolder).searchManagedEntities(EntityType.POOL.getValue()))
+                                 .stream()
+                                 .map(resourcePool -> (ResourcePool) resourcePool)
+                                 .collect(Collectors.toCollection(ArrayList::new));
+        } catch (RemoteException e) {
```

Comment:
And is it actually better? I did it on purpose as it seems there is no guarantee on the List returned by `toList()`, look at the first reply here: https://stackoverflow.com/questions/21912314/what-kind-of-liste-does-collectors-tolist-return